pkg/relay/ethereum/tx.go

```diff
@@ -48,49 +48,127 @@ func (c *Chain) SendMsgs(msgs []sdk.Msg) ([]core.MsgID, error) {
 		opts.NoSend = true
 		tx, err = c.SendTx(opts, msg, skipUpdateClientCommitment)
 		if err != nil {
-			logger.Error("failed to send msg / NoSend: true", err, "msg_index", i)
+			logger.Error("failed to build tx for gas estimation", err, logAttrMsgIndex, i)
 			return nil, err
 		}
+		var rawTxData string
+		if bz, err := tx.MarshalBinary(); err != nil {
+			logger.Error("failed to encode tx", err, logAttrMsgIndex, i)
+		} else {
+			rawTxData = hex.EncodeToString(bz)
+		}
 		estimatedGas, err := c.client.EstimateGasFromTx(ctx, tx)
 		if err != nil {
-			revertReason, rawErrorData, err2 := c.getRevertReasonFromEstimateGas(err)
-			if err2 != nil {
-				logger.Error("failed to get revert reason", err2, "msg_index", i)
+			var revertReason, rawErrorData string
+			if reason, data, err := c.getRevertReasonFromEstimateGas(err); err != nil {
+				// Raw error data may be available even if revert reason isn't available.
+				rawErrorData = hex.EncodeToString(data)
+				logger.Error("failed to get revert reason", err,
+					logAttrRawErrorData, rawErrorData,
+					logAttrMsgIndex, i,
+					logAttrRawTxData, rawTxData,
+				)
+			} else {
+				revertReason = reason
+				rawErrorData = hex.EncodeToString(data)
 			}
 
-			logger.Error("failed to estimate gas", err, "revert_reason", revertReason, "raw_error_data", hex.EncodeToString(rawErrorData), "msg_index", i)
+			logger.Error("failed to estimate gas", err,
+				logAttrRevertReason, revertReason,
+				logAttrRawErrorData, rawErrorData,
+				logAttrMsgIndex, i,
+				logAttrRawTxData, rawTxData,
+			)
 			return nil, err
 		}
 		txGasLimit := estimatedGas * c.Config().GasEstimateRate.Numerator / c.Config().GasEstimateRate.Denominator
 		if txGasLimit > c.Config().MaxGasLimit {
-			logger.Warn("estimated gas exceeds max gas limit", "estimated_gas", txGasLimit, "max_gas_limit", c.Config().MaxGasLimit, "msg_index", i)
+			logger.Warn("estimated gas exceeds max gas limit",
+				logAttrEstimatedGas, txGasLimit,
+				logAttrMaxGasLimit, c.Config().MaxGasLimit,
+				logAttrMsgIndex, i,
+				logAttrRawTxData, rawTxData,
+			)
 			txGasLimit = c.Config().MaxGasLimit
 		}
 		opts.GasLimit = txGasLimit
 		opts.NoSend = false
 		tx, err = c.SendTx(opts, msg, skipUpdateClientCommitment)
+		if bz, err := tx.MarshalBinary(); err != nil {
```

Comment:
@siburu  If the `tx` is nil(and err !=nil)  at line 96, doest it cause a nil reference error?

Reply:
@bluele I'm very sorry. I have fixed it. https://github.com/datachainlab/ethereum-ibc-relay-chain/pull/44/commits/5178fa5a702cbd011e83c4f56bda378cd71db8c7